### PR TITLE
Add global discovery

### PR DESCRIPTION
**Pull request by msdemlei**
This PR should eventually add a facility for global dataset discovery to pyVO.  Specifically, people should be able to say "give me all images in the VO showing M42 around 42nm on MJD 54125."

Realistically, we ought to allow intervals on input and probably multiple positions (though we should refuse SIA1 for multiple objects).

At the moment, none of this will work; I'm missing two very basic building
blocks that I won't have time to locate before I'm going on holiday, both related to the way pyvo treats query results:

* I need to remove rows from registry query results.  That's nasty because RegistryResults reads from the astropy.VOTable instance, and that doesn't support dropping rows so far for all I can see.
* I need to turn whatever comes back from SIA1 and TAP services to sia2.ObsCoreRecord-s.  Again, as implemented so far, this wants a VOTable to pick stuff from.

If there is no better solution for both, we probably need to turn things into astropy tables, manipulate them, and then turn them back to VOTables.  It'd be great if there were a less convoluted way to do this.  Is there?

Even if that's there, this won't be ready for prime time; we will at least need defined timeouts on all services involved, because there's always *some* services in the VO hanging clients for almost arbitrary times.

Once we have the image case, a similar modules for spectrum discovery (in SSA and Obscore for now) should be simple.

We probably should talk to the IVOA to make throwing out capabilities leading to the same data possible/more robust; I think we ought to suggest that SIAv2 services over data that's also there in Obscore services should have an IsDerivedFrom set to the respective TAP service (or Obscore resource).

So, there's a metric ton of TODOs – I'm grateful for commits fixing any of them.

Ah – and before you ask: Testing code like this that talks to services all over the internet is of course a particular challenge; I think I have a plan, so don't worry about that particular issue too much at this point.

### CODE REVIEW

File: docs/utils/testing.rst [hunk 1]
```diff
@@ -0,0 +1,50 @@
+.. _pyvo-testing:
+
+******************************************
+Helpers for Testing (`pyvo.utils.testing`)
```
Comment (bsipocz) [inline docs/utils/testing.rst:4]: bikeshed  at this point - not sure how to do this best, but I feel these should be private functions (aka for consumptions only within pyvo), yet I agree that having API docs for them would be nice rather than hiding away.

maybe just an extra sentence that say that?

**Comment by msdemlei** [comment]
So... I think this could now be worth a look.  See also my blog post on this: https://blog.g-vo.org/global-dataset-discovery-in-pyvo.html.

The CI failures look interesting – I half fear they are related to implementation details changing the request hashes that utils.testing generates.  Perhaps we'll have to properly scaffold these things after all?  Anyway, I'll look at this some other day; it shouldn't impact discovery as such.

Meanwhile, on the usual codestyle failures I have to say I'm fairly unhappy that I'll now have to say foo / 2 rather than foo/2 – does that really help someone?  And do people here in pyVO actually prefer the former?

If we have to have these checks in place: Can someone recommend a formatter/pretty-printer that I can run halfway blindly on the sources?  Frankly, I don't think I'd like to get used to typing foo / 2 or 1 + 2, so if this kind of thing can happen in a pre-commit hook, that'd make my life easier...

**Comment by bsipocz** [comment]
> f we have to have these checks in place: Can someone recommend a formatter/pretty-printer that I can run halfway blindly on the sources? Frankly, I don't think I'd like to get used to typing foo / 2 or 1 + 2, so if this kind of thing can happen in a pre-commit hook, that'd make my life easier...

even old basic tools could fix this, autopep8, or I have flymake set up on my emacs to check on the codestyle (and it should pick up the setting from the repo)

**Comment by msdemlei** [comment]
So... I think modulo review points and perhaps a few style matters, this could be merged, with a caveat that the API is probably not stable yet.  It does something interesting, and although data providers still can do quite a bit to make this work better, it's a lot more likely they will do that if there is merged code in pyVO.

From my perspective, the main problem is testing.  Since there is so much network activity going on here, my plan has been to ship frozen network answers as part of the source (see utils.testing for the general scheme).

It turns out that that is nowhere near as simple as I had hoped.  The test failures with old versions of python are almost certainly due to some subtle difference in the serialisation of query parameters; they might be fixable by normalising these a bit more aggressively, but perhaps they are not; I'd have to pull up systems with these old versions to look.

But: Is this the right way to go about this?  Perhaps the instrumentation should sit in TAPService, where the parameters are much less dependent on versions than where requests.requests has written its payload?  Or should we, especially after the xz disaster, try harder to avoid what looks like binary blobs in the VCS in the first place?  But then how do we do the scaffolding for these tests without going insane?

Thoughts and/or ideas?

**Comment by ManonMarchand** [comment]
Hi Markus, 
With which python version do the tests pass without access to remote data? Maybe there was a change in the way hashs are calculated that we could identify in python changelogs?

**Comment by msdemlei** [comment]
Ah well... failures up and down.  I'll dig a bit deeper tomorrow, but this does look as if we're going to need a giant scaffold.  Oh dang.

**Comment by msdemlei** [comment]
Oh yikes.  I've tried this on two different Debian relases (so, I'm not reproducing github's environment).  The first instance for differing hashes is because astropy includes something like "Produced with astropy.io.votable version 5.2.1" in its VOTables (which get uploaded in the course of the tests).  Updating the cached files depending on the underlying astropy release obviously is not an option.

I think the testing plan with intercepting and caching https requests and responses and recovering the responses based on hashes of the requests is fundamentally broken for something as complex as the global discovery.

I suppose we'll have to mock at a higher level.  If someone has good ideas how to go about this: I'd be most grateful.

**Comment by bsipocz** [comment]
> I suppose we'll have to mock at a higher level. If someone has good ideas how to go about this: I'd be most grateful.

I don't have any suggestions without diving into this properly, but will try to find time to do so next week.

**Comment by msdemlei** [comment]
After some deliberation I have decided that doing local tests properly for many of the features is really hard (and would hide a lot of interesting bugs).  Hence, I have, with a solid helping of regret, made the more interesting tests remote.  They're still relatively quick.

So... Volunteers for reviewing this?  It'd be lovely to have it merged into main for improved visibility, even though there will be quite a bit of refactoring once we also do global spectrum discovery.

The codecov analysis is way off (I suspect they don't do remote tests).  In my local pytest --cov, I only have 14 uncovered lines, and most of these are odd-ish (well, the obscore vs. tap-obscore thing should eventually be covered, but that's not easy and I hope this code will be obsolete rather soon anyway).  

The doctest failures... I think that's a separate matter independent of this PR.  We really should think about dropping a few of them or at least figure out why the doctests take so long in the first place.

Documentation building has some fairly odd messages, some of which may be introduced with the PR.  I'd appreciate if someone with more automodapi-fu than me could have a look at that and figure out how to silence/fix these diagnostics.

**Comment by bsipocz** [comment]
> So... Volunteers for reviewing this? It'd be lovely to have it merged into main for improved visibility, even though there will be quite a bit of refactoring once we also do global spectrum discovery.


I know that there are some disagreements about this approach, maybe we should clear that up first next week, then I can put this on my list to review.


> The codecov analysis is way off (I suspect they don't do remote tests). In my local pytest --cov, I only have 14 uncovered lines, and most of these are odd-ish (well, the obscore vs. tap-obscore thing should eventually be covered, but that's not easy and I hope this code will be obsolete rather soon anyway).

I'll check this once we do the code review, but recently there were some codecov issues elsewhere, too, so it may very well be a red herring.

 
> The doctest failures... I think that's a separate matter independent of this PR. We really should think about dropping a few of them or at least figure out why the doctests take so long in the first place.

separate, I'm opening the fix for it.

> Documentation building has some fairly odd messages, some of which may be introduced with the PR. I'd appreciate if someone with more automodapi-fu than me could have a look at that and figure out how to silence/fix these diagnostics.

I volunteer for fixing these, once we go ahead with the code review.

File: pyvo/registry/regtap.py
Comment (bsipocz) [inline pyvo/registry/regtap.py:1]: These changes make a lot of sense in a separate bugfix PR, too.

File: pyvo/registry/rtcons.py
Comment (bsipocz) [inline pyvo/registry/rtcons.py:1]: These changes would make sense without the discovery module, too. Could they be added in a separate PR maybe?

File: pyvo/dal/sia.py
Comment (bsipocz) [inline pyvo/dal/sia.py:1]: This are unrelated fixes, would be nice to have them as a separate small PR.

**Review by bsipocz — COMMENTED** [review]
Some things seems to be unrelated but useful bugfixes. I wonder if they can be done separately, thus we could include them in a release in a more timely manner?

**Comment by bsipocz** [comment]
@trjaffe - Could you, and/or someone from the registry group, have a look at this? 
I recall there were concerns, but the details (affects on this PR) are not super clear to us maintainers, so it would be great to get some feedback before we jump into the code review ourselves.

**Comment by msdemlei** [comment]
On Fri, Jun 14, 2024 at 03:53:12PM -0700, Brigitta Sipőcz wrote:
> Some things seems to be unrelated but useful bugfixes. I wonder if
> they can be done separately, thus we could include them in a
> release quicker?

They could, but to be very frank I would much rather see the whole
thing merged for 1.6 so we have at least *something* in terms of
global dataset discovery in there.  This has been lingering far too
long (and that after I've procrastinated starting it for much longer
even).

If there are serious concerns with the main discovery code that I
cannot address quickly, I'd be ok with pulling out the bugfix parts;
but they are close enough to the issue of global dataset discovery
that I don't think there's a strong logical reason to pull them out.

**Comment by trjaffe** [comment]
The NAVO Python working group has been looking at PR 470 and have a few questions. (TJ is just speaking for the team.)  

We hope to see it merged soon, though we think it needs a few tweaks like marking the module as a prototype first. A few of us are looking at it, but I also have these questions for you:

1. Can you expand on how to use the methods that make up discovery.images_globally()? I mean specifically when the user wants to do some kind of detailed registry search and then hand the resulting list of services. I tried the code shown below and it gives me a TypeError, so I probably have something simple wrong. 
```
import pyvo
from astropy import units as u
from astropy import time
from pyvo import discover 
services = pyvo.registry.search(keywords=["heasarc rass"]) 
im_discoverer = discover.image.ImageDiscoverer(space=(274.6880, -13.7920, 0.1))
im_discoverer.set_services(services) 
results, log = im_discoverer.query_services()
```

 2. The example given in the docs, the very first code cell on https://docs.g-vo.org/temp-pyvo-global-discovery/discover/index.html, returns no results for me. Is this expected? It runs and the log I get says:
```
['Skipping ivo://org.gavo.dc/__system__/siap2/sitewide because it is served by ivo://org.gavo.dc/__system__/obscore/obscore',
 'Obscore GAVO Data Center Obscore Table: 0 records']
```
We should make sure the examples return something useful.

3.  Just to confirm, I think the other issue I ran into is something you mentioned in your blog post that is currently unsolved: doing a registry search, making some selection among the records, and then handing that edited list to the discoverer. I also show what I want to be able to do in the example code below. Let me know if you know of a way to do this. It’s part of a broader awkwardness handling these PyVO objects that are like lists but perhaps not enough like lists.  This probably needs a separate issue and PR to address.
```
records = []
for s in services:
  # Doesn't matter what this is, just make a selection of a subset of services:  
  if "rosat" in s.res_description.lower() and "rass" in s.res_description.lower():
    records.append(s)
# How to create a RegistryResults object from a list of RegistryResource objects? 
selected_services = pyvo.registry.regtap.RegistryResults(records)
im_discoverer.set_services(selected_services)
```

**Comment by msdemlei** [comment]
On Fri, Jul 19, 2024 at 09:26:20AM -0700, trjaffe wrote:
> In re the prototype, yes, that was the suggestion, though not
> because the standards themselves are not approved yet.  It's more
> that this is a complicated workflow and we want to spend some time
> testing it and see how it goes reworking some of our tutorials
> based on it before it becomes 'official'.  And realistically
> speaking, that will take us some more time.

Hm.  I have tried for another 10 minutes to like pyvo.utils.prototype
and didn't manage.  Frankly, I probably don't even like it for what
it was designed for (enabling code relying on remote features that
may or may not be present), but that may be because it solve a
problem I have never had.

But for what we have here, just telling users "if you use this, be
aware that you may have to adapt your program later because we're
still designing the API", I don't think it works at all: when people
type

  pyvo.utils.activate_features("global-discovery")

because they're seeing it in some example, they will fairly certainly
not suspect that means `when you type "from pyvo import discover",
you're using an unstable API'.  When they type it because they read
the documentation and know what it means, they've already *read* that
the API may be unstable and don't need to spoil their code with
function calls that will appear fairly cryptic later.

So... I have fairly bad aches when contemplating the use prototype
here.  On my pain level scale
<https://blog.g-vo.org/building-consensus.html#scale>, I'd give it an
7 going towards 8.

Hm: perhaps I can accomodate your concerns if I just emit a warning
("Preprecation", if you will -- I'm surprised that apparently almost
no one else has come up with that term yet) to the effect that this API
is unstable on importing discover? At least for now, "import pyvo"
doesn't pull in discover, and so you would exactly hit the people
trying to use the new API, and you could exactly say why you're
warning.

Deal?

**Comment by tomdonaldson** [comment]
>Hm: perhaps I can accomodate your concerns if I just emit a warning ("Preprecation", if you will -- I'm surprised that apparently almost no one else has come up with that term yet) to the effect that this API is unstable on importing discover? At least for now, "import pyvo" doesn't pull in discover, and so you would exactly hit the people trying to use the new API, and you could exactly say why you're warning. 
> 
>Deal?

I'm inclined to go along with this.  I don't have the same issues with the prototype construct, and would be comfortable using the prototype feature in this circumstance, but I do think the key issue is to somehow make a user aware that this API may evolve rather quickly without necessarily undergoing normal deprecation cycles.

Indeed I think it is important to get this merged to get more people using it and contributing feedback.  My gut feeling is that there will be aspects that work well and those that don't and the more usage we can encourage the more improvement can be made.  Moving forward I hope we can dramatically reduce the amount of time and overhead in getting changes merged as that will both facilitate and encourage faster evolution of code base.

@msdemlei would you be able to take a pass at resolving the conflicts, fixing the tests, and introducing a warning message of your choice either in the code or documentation?  Let us know if we can help with the mechanics of any of that.  Then we can do a last Python-focused review and get this merged.

**Comment by bsipocz** [comment]
> Aw, if only it were so simple.

I'm happy to make the warning work with the CI system if that is the decision we go ahead with instead of the prototype route.

**Comment by bsipocz** [comment]
(fyi: we don't use pytest.ini at all, all the pyetst configs are in `setup.cfg`, but again, to rehash, I'm happy to make the warning in a way that works with our testing etc, once it's clear what type of warning/messaging is decided upon).

**Comment by tomdonaldson** [comment]
I'm very comfortable with the proposed text of the warning (`pyvo.discover's API is still under design in pyVO 1.6 and may change without prior notice.  Feedback to the authors is most welcome.`).

I'm not familiar enough with the possibilities to weigh in on the best place to trigger the warning, but would be happy with anything @bsipocz proposes.

**Comment by andamian** [comment]
I haven't thought about this before, but another alternative to the pesky warning is to do a pre-release in `pypi` for this feature. It's easy to install for whomever wants to try it. The two main consequences with this is that 1. this should be limited in time to a few months up to a year and 2. users need to be informed to try the feature through other means - they will not just stumble on it.
That being said, I'm fine with the warning too.

**Comment by bsipocz** [comment]
> Re pypi pre-release

I expect this to be released much sooner, and the API to be still open to changes after 1.6 is out. (But if anyone is expected to run into installation issues doing a pip install from github, then we can certainly push a pre-release to pypi).

**Comment by duytnguyendtn** [comment]
> The trouble is that if I raise the warning in discover's __init__.py
(which I think makes the most sense and has the advantage of being
dead simple), pytest's run fails *during test collection*

Could explicitly request pytest to check for the warning work here? I'm thinking something like https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions

```
import pytest


def test_zero_division():
    with pytest.raises(ZeroDivisionError):
        1 / 0
```

but instead, check for the warning you're raising here. If the warning is coming up on import, you might need to wrap the import in the `with pytest.raises` call.

(PS Hi! I'm a dev working under TJ; I'm doing a quick deep dive to see if I can help with the table dropping/appending issue)

**Comment by duytnguyendtn** [comment]
Hi @msdemlei!

> I need to remove rows from registry query results. That's nasty because RegistryResults reads from the astropy.VOTable instance, and that doesn't support dropping rows so far for all I can see.

I was playing around with the `DALResultsTable` and I think I might have found a proof of concept that deletion is at least possible:

image

Right now, it's quite crude, but after a little of refining and some convenience methods, I think it could allow what you're asking for. Let me know your thoughts! Also @bsipocz your opinion would be appreciated here too, before I get too deep into implementation!

**Comment by msdemlei** [comment]
On Thu, Aug 29, 2024 at 01:41:37PM -0700, Duy Tuong Nguyen wrote:
> Right now, it's quite crude, but after a little of refining and
> some convenience methods, I think it could allow what you're asking
> for. Let me know your thoughts! Also @bsipocz your opinion would be
> appreciated here too, before I get too deep into implementation!

Well... for the present PR I don't think it would help a lot; I'm
quite happy to wrap the services up in Queriable-s, and these
wouldn't fit into RegistryResults anyway.

The more general thing, letting people delete from DALResults...
Well, I'm sure there are situations when people would be grateful if
they were able to do that.  Which means I'm mildly in favour of
investigating what it would take.  On the other hand, mutability is a
curse, and we'd have to have a close look at the various classes
inheriting from DALResults; if any of them kept indexes, that would
be a severe blow to the concept of mutating .array from underneath
the VOTable.

Also: Do we know the runtime properties of get_first_table().array?
If this does anything non-trivial, then we definitely should not
implement things in this way; once we offer deletion, people will, I
think, expect O(1) runtime in the typical case (an occasional O(n) to
pack things is ok, but you don't want an individual deletion to take
measurable time).

In short: I'd step carefully :-)

**Comment by msdemlei** [comment]
On Wed, Sep 04, 2024 at 07:08:20AM -0700, Duy Tuong Nguyen wrote:
> largest operation would be making a copy of the new array, an
> unfortunate side-effect of the immutability of numpy ndarrays.

Yes, and as I commented over on the astropy PR, I *think* that's by
and large a killer; there's a reason why you cannot just pop rows out
of numpy ndarrays, and I *think* that reason applies to us, too.

> generic case, and this would also force downstream packages to
> explicitly declare whether their tables are meant to mutable or
> not. In those cases, they can override the method to disable the
> user's ability to delete from the `TableElement`. I think this
> exercise of explicitly declaring mutability or immutability for

I agree that being explicit here would be a great thing.  And also
to provide clear guidance on "Ok, so you have a VOTable, and you want
to create a VOTable with more or fewer rows.  Here's what to do."
Having this for the one-table case first would already be a good
thing, paying attention to keeping any additional RESOURCE elements,
though (e.g., datalink, perhaps mivot).

But thanks a lot for investigating this.  It's always good to get a
reality check on one's wishes :-).

File: docs/discover/index.rst [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+  from pyvo import discover
+  from astropy import units as u
+  from astropy import time
+
+  datasets, log = discover.images_globally(
+    space=(273.5, -12.1, 0.1),
+    spectrum=1*u.nm,
+    time=(time.Time('1995-01-01'), time.Time('1995-12-31')))
+  print(datasets)
```
Comment (bsipocz) [inline docs/discover/index.rst:55]: Is this expected to run long and therefore not doctested? This question stands for all the code examples in this file.
Even if they need to be skipped for a reason I would rather do it the same way we do skip modules with doctestplus, on the file level, rather than not formatting them for doctest to collect. I'm happy to fix up the file, just let me know the thought process

File: pyvo/discover/tests/test_imagediscovery.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    with pytest.raises(dal.DALQueryError) as excinfo:
+        image.ImageDiscoverer().query_services()
+    assert "No services to query." in str(excinfo.value)
+
+
+# Tests requiring remote data below this line
+
+
+@pytest.mark.remote_data
+def test_single_sia1():
```
Comment (bsipocz) [inline pyvo/discover/tests/test_imagediscovery.py:192]: This test is failing for me with 

```
E           pyvo.dal.exceptions.DALQueryError: No services to query.  Unless you overrode service selection, you will have to loosen your constraints.
```

File: docs/index.rst [hunk 4]
```diff
@@ -132,9 +132,11 @@ Using ``pyvo``
...
    dal/index
    registry/index
+   discover/index
    io/index
    auth/index
    samp
    mivot/index
    utils/index
    utils/prototypes
+   utils/testing
```
Comment (bsipocz) [inline docs/index.rst:142]: there is no `docs/utils/testing.rst` file, so this causes an error.

File: pyvo/dal/sia.py
Comment (bsipocz) [inline pyvo/dal/sia.py:1]: I suspect these changes can be factored out. I just make a note of it and see how quickly we we merge everything from here.

File: docs/discover/index.rst [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  im_discoverer.query_services()
+  print(im_discoverer.log_messages)
+  print(im_discoverer.results)
+
+
+
+Reference/API
+=============
+
+.. automodapi:: pyvo.discover.image
```
Comment (bsipocz) [inline docs/discover/index.rst:213]: image is not a public namespace, expose only the public API from the module into the user docs
```suggestion
.. automodapi:: pyvo.discover
```

File: pyvo/discover/__init__.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+import warnings
+from pyvo.utils.prototype import PrototypeWarning
+
+# if you remove this warning, also remove the ignorere in test_imagediscovery.
+warnings.warn("pyvo.discover's API is still under design in pyVO 1.6 and"
+    " may change without prior notice.  Feedback to the authors is most"
+    " welcome.", PrototypeWarning)
+
+from .image import images_globally, ImageDiscoverer
```
Comment (bsipocz) [inline pyvo/discover/__init__.py:14]: Make public API explicit, e.g. don't expose `warnings`, etc.
```suggestion
from .image import images_globally, ImageDiscoverer


__all__ = ['images_globally', 'ImageDiscoverer']
```

File: pyvo/discover/image.py [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+        way.
+    spectrum :
+        An astropy quantity convertible to a (vacuum) wavelength; images
+        must cover this point in the (electromagnetic) spectrum.
+    time :
+        An astropy time that must be in the observation time of the image
+        (if it declares a time).
+    inclusive :
+        Set to True to incluse services that do not declare their
+        STC coverage.  By 2023, it's a good idea to do that as many
```
Comment (bsipocz) [inline pyvo/discover/image.py:602]: remove year

File: pyvo/discover/image.py [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+        relevant archives do not do that.
+    watcher :
+        A callable that will be called with the ImageDiscoverer instance and
+        a string perhaps suitable for displaying to a human.
+    services :
+        An optional `~pyvo.registry.RegistryResults` instance to
```
Comment (bsipocz) [inline pyvo/discover/image.py:608]: use the public namespace (we may want to clean these up and have `RegistryResults` documented in registry rather than in registry.regtap, I'm sure there are a couple of related issues already)
```suggestion
        An optional `~pyvo.registry.regtap.RegistryResults` instance to
```

File: pyvo/discover/image.py [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+        override automatic services detection.
+
+    When an image has insufficient metadata to evaluate a constraint, it
+    is excluded; this mimics the behaviour of SQL engines that consider
+    comparisons with NULL-s false.
+
+    Returns
+    -------
+    discovered_images, log_lines
+        The images found are returned in a list of `ImageFound` instances.
```
Comment (bsipocz) [inline pyvo/discover/image.py:618]: Not in the public API
```suggestion
        The images found are returned in a list of ``ImageFound`` instances.
```

File: docs/discover/index.rst [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+  from astropy import time
+
+  datasets, log = discover.images_globally(
+    space=(273.5, -12.1, 0.1),
+    spectrum=1*u.nm,
+    time=(time.Time('1995-01-01'), time.Time('1995-12-31')))
+  print(datasets)
+
+The function returns a pair of lists.  ``datasets`` is a list of
+`~pyvo.discover.image.ImageFound` instances.  This is the (potentially
```
Comment (bsipocz) [inline docs/discover/index.rst:58]: Is this supposed to be public API? If yes, is the implementation detail of the `image` important, or can we flatten things out and expose it in `discover`?

File: docs/discover/index.rst [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+If in doubt, have a brief look at the log lines; if a service that
+sounds promising shows a timeout, perhaps try again with a longer
+timeout or use partial matching.
+
+
+Overriding service selection
+----------------------------
+
+You can also pass a `pyvo.registry.RegistryResults` instance to
```
Comment (bsipocz) [inline docs/discover/index.rst:148]: namespace
```suggestion
You can also pass a `pyvo.registry.regtap.RegistryResults` instance to
```

File: docs/discover/index.rst [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+For finer control of the discovery process, you can directly use
+the `pyvo.discover.image.ImageDiscoverer` class.  It is constructed with
+essentially the same parameters as the search function.
+
+To run the discovery, first establish which services to query.  There
+are two ways to do that:
+
+* Call the ``discover_services()`` method.  This is what the search
+  function does; it uses your constraints as above.
+* Pass a `pyvo.registry.RegistryResults` instance to ``set_services``.
```
Comment (bsipocz) [inline docs/discover/index.rst:165]: namespace
```suggestion
* Pass a `pyvo.registry.regtap.RegistryResults` instance to ``set_services``.
```

File: pyvo/discover/image.py [hunk 13]
```diff
@@ -0,0 +1,635 @@
...
+from .. import dal
+from .. import registry
+from ..registry import regtap
+
+
+# imports for type hints
+from typing import Callable, Generator, List, Optional, Set, Tuple
+from astropy.units import quantity
+
+
```
Comment (bsipocz) [inline pyvo/discover/image.py:42]: If anything else not imported into ``__init__.py`` of the module an exposed as public API from there, it should be added here in ``__all__`` and exposed as such. The rest cannot be referenced in the API docs as a sphinx link.

I would think that a flatter namespace would be nice, so would advocate for exposing anything public though the module, and keep this file as an implementation detail, but I may miss some reasons to do it otherwise. Either case, we need to be explicit.

File: pyvo/registry/__init__.py [hunk 14]
```diff
@@ -5,7 +5,9 @@
 The regtap module supports access to the IVOA Registries
 """
 
-from .regtap import search, ivoid2service, get_RegTAP_query, choose_RegTAP_service
+from .regtap import (search, ivoid2service, get_RegTAP_query,
+    choose_RegTAP_service,
+    RegistryResults, RegistryResource)
```
Comment (bsipocz) [inline pyvo/registry/__init__.py:10]: Is this change intentional as part of this PR? If yes, what's the purpose?

(If the intent is to consolidate the public API and remove the `regtap` implementation detail, I would strongly suggest to do it separately from this PR.

File: pyvo/registry/rtcons.py
Comment (bsipocz) [inline pyvo/registry/rtcons.py:1]: Can I assume these are unrelated and therefore can be factored out?

File: pyvo/registry/tests/test_rtcons.py
Comment (bsipocz) [inline pyvo/registry/tests/test_rtcons.py:1]: are these related?

**Review by bsipocz — COMMENTED** [review]
I'm sorry this review took so long. I pushed a few commits that are fixing the pytest collections, and the unrelated looking tests that were failing locally.

I have also left a lot of maintenance related comments, a lot of them are comments about how to make the documentation build. I have all those changes locally with a passing build, so I'm happy to push them as a commit if you prefer.

I also run into the concern about the namespace and what will eventually end up exposed to the users. I would advocate for being explicit, and flatten the namespace, and left comments about it.

I also noted a couple of unrelated looking changes. I haven't looked into the details to see if those are due to changes that happened on `main` since this one is being open and are rebase remnants, or can be factored out into their own bugfix PRs. If the latter I'm happy to do the shuffling around.

That all being said, I haven't tried to use this all, I believe all that has been covered in previous rounds of reviews. So my concerns here were limited to get a passing docs build and passing tests.

As for the tests, I see one remaining relevant one. And have no idea what goes wrong with mivot, those should be absolutely unrelated and a red herring here.

File: docs/discover/index.rst (hunk 2 above)
Comment (msdemlei) [reply docs/discover/index.rst:55]: Yes, although the requested coverage probably will not match many services in the future, I think this is too unpredictable to work as part of anything that runs on every push.  So, I'd rather skip it, yes.

File: pyvo/discover/tests/test_imagediscovery.py (hunk 3 above)
Comment (msdemlei) [reply pyvo/discover/tests/test_imagediscovery.py:192]: Ahem.  I accidentally unpublished that service as version 2 of that resource came in.  That particular service will remain, though, and I've re-published it.  The test succeeds again.

File: docs/index.rst (hunk 4 above)
Comment (msdemlei) [reply docs/index.rst:142]: Sorry about this; there *was* a testing.rst (about building scaffolds by observing HTTP requests, which fails because of version strings and such) and I forgot to remove that reference when I dropped it for this PR (I'll re-introduce testing helpers in another PR, though, but that's different material).

File: docs/discover/index.rst (hunk 5 above)
Comment (msdemlei) [reply docs/discover/index.rst:213]: Ok, thanks.

File: pyvo/discover/image.py (hunk 7 above)
Comment (msdemlei) [reply pyvo/discover/image.py:602]: Hm... I have updated to 2024, and I'm saying "as of" now, but I do hope that coverage declaration will become much more common once software like this is used in the wild, so I'd like to give an idea of a horizon.

File: docs/discover/index.rst (hunk 10 above)
Comment (msdemlei) [reply docs/discover/index.rst:58]: People are supposed to look inside ImageFound-s, and hence I suppose it needs to be public API.  I have changed as much.

File: pyvo/registry/rtcons.py
Comment (msdemlei) [reply pyvo/registry/rtcons.py:1]: Well... obscore-new comes from TableReg (https://ivoa.net/documents/Notes/TableReg/), which is really an integral part of this effort, and the "inclusive" options are really necessary here, too, but probably not in many other contexts.  So, I'd say these API extensions are in a sufficiently narrow context of global discovery to keep them in here.  But then I admit that assessment may be influenced by my lazyness.

File: pyvo/registry/tests/test_rtcons.py
Comment (msdemlei) [reply pyvo/registry/tests/test_rtcons.py:1]: Yes, in the sense of the last answer.

**Comment by msdemlei** [comment]
Ok -- I've factored out the registry API change to PR #608.  Feel free to rebase, and if you want to pull out the inclusive feature, I'll certainly not stand in your way, but I'd consider these a lot less critical for separate visibility.

File: docs/discover/index.rst (hunk 2 above)
Comment (bsipocz) [reply docs/discover/index.rst:55]: Ok, so I'll make a note for the infrastructure side that we need a directive for slow doctests (we have such markers for the normal tests), and then can just run the slow ones in the weekly cron

**Comment by bsipocz** [comment]
Yes, mivot failings feels to be unrelated, so I'm fixing the docs build one more time and this is good to go.

**Review by bsipocz — APPROVED** [review]
I'll go ahead and merge this now, in agreement with Tom and Adrian.

If the mivot failures are persistent, we'll deal with them in a follow-up.

**Comment by bsipocz** [comment]
Thank you so much for your patience here, Markus!
